File: frontend/src/features/assistants/model/tools.ts
/**
 * The tool allow-list source for the assistant builder (#212 → #505, ADR-0011 §1).
 *
 * The candidate tools now come from the SERVER (`GET /tools`, the member-readable
 * tool catalogue added in #505), not from a hardcoded stub. The old `KNOWN_TOOLS`
 * constant listed four retrieval tools and nothing else, so `run_python` — a real,
 * registered, T2 tool the backend has always accepted in `toolAllowlist` — was
 * simply unreachable from the UI: the only assistant that had it was seeded straight
 * into the database.
 *
 * What stays local is pure PRESENTATION: a friendly label per known tool name, and
 * the risk copy that makes a side-effecting tool legible. Governance is never
 * decided here — `enabled` / `requires_approval` / `risk_tier` all arrive from the
 * catalogue, which reflects the tenant's admin policy.
 *
 * An empty allow-list on the wire (`[]`) still means "the default retrieval tool
 * set" (per the contract), so an assistant with nothing ticked is still useful.
 */
import type { RiskTierId, ToolCatalogEntry } from '@/api';

/** One selectable tool: the catalogue entry plus the copy the checkbox renders. */
export interface ToolOption {
  /** The wire name persisted into `toolAllowlist`. */
  name: string;
  label: string;
  description: string;
  riskTier: RiskTierId;
  /** False for a side-effecting (write-tier) tool — surfaced, never hidden. */
  readOnly: boolean;
  /**
   * False when the tenant's admin has switched the tool off. Such a tool is shown
   * as UNAVAILABLE rather than offered as a control the run-time gate would refuse.
   */
  enabled: boolean;
  /** True when an invocation is still approval-gated for this tenant. */
  requiresApproval: boolean;
  /**
   * False for a name already in the allow-list that the catalogue does not describe
   * (a retired registry tool, or an `mcp:*` tool granted elsewhere — those are
   * validated server-side against the tenant's MCP servers, so they are LEGITIMATE
   * grants, not junk). Such a row has NO authoritative governance metadata, so the
   * other fields on it are placeholders and must not be presented as facts: it gets
   * its own conservative copy rather than borrowing the admin-disabled or T0
   * read-only presentation, either of which would state something untrue.
   */
  catalogued: boolean;
}

/**
 * Friendly labels for the tools we ship. A name with no entry here falls back to its
 * wire name, so a newly registered backend tool appears in the picker immediately
 * (with its registry description) instead of being invisible until the UI catches up
 * — the failure mode #505 is about.
 */
const LABELS: Record<string, string> = {
  search_text: 'Text search',
  search_documents: 'Document search',
  list_documents: 'List documents',
  get_document: 'Read document',
  web_search: 'Web search',
  write_file: 'Write a file',
  run_python: 'Run Python (code execution)',
};

/** A human label for a tool name; falls back to the wire name itself. */
export function toolLabel(name: string): string {
  return LABELS[name] ?? name;
}

/** Project a catalogue entry into the option the picker renders. */
export function toToolOption(entry: ToolCatalogEntry): ToolOption {
  return {
    name: entry.tool_name,
    label: toolLabel(entry.tool_name),
    description: entry.description,
    riskTier: entry.risk_tier,
    readOnly: entry.read_only,
    enabled: entry.enabled,
    requiresApproval: entry.requires_approval,
    catalogued: true,
  };
}

/**
 * The picker's rows: every catalogue entry, plus a placeholder for any name already
 * in the allow-list that the catalogue does not describe (a tool retired from the
 * registry, or an `mcp:*` tool granted elsewhere). Such a row is rendered so the
 * owner can SEE and remove it — dropping it silently would rewrite their config
 * behind their back on the next save.
 */
export function toToolOptions(entries: ToolCatalogEntry[], selected: string[]): ToolOption[] {
  const options = entries.map(toToolOption);
  const known = new Set(options.map((o) => o.name));
  const orphans = selected
    .filter((name) => !known.has(name))
    .map<ToolOption>((name) => ({
      name,
      label: toolLabel(name),
      description:
        'Already granted, but not described by this tenant’s tool catalogue. Its risk and' +
        ' governance are unknown here, so it is shown conservatively.',
      // Placeholders ONLY — `catalogued: false` is what the renderer branches on.
      // Deliberately NOT `T0` + `readOnly: true` + `enabled: false`: that trio made an
      // `mcp:*` grant render as a harmless read-only tool that "an admin turned off"
      // (no admin did), suppressing the side-effecting warning on a component whose
      // contract is "governance is displayed, never bypassed". `readOnly: false` keeps
      // the conservative side-effecting copy; `enabled: true` keeps the row unlocked so
      // a still-granted tool can be re-ticked after an accidental untick.
      riskTier: 'T0',
      readOnly: false,
      enabled: true,
      requiresApproval: false,
      catalogued: false,
    }));
  return [...options, ...orphans];
}

File: backend/app/api/v1/tools.py
"""Tool-catalogue route — ``GET /tools`` (issue #505).

The **member-readable, read-only** projection of the per-tenant tool governance the
assistant builder needs to offer a truthful allow-list picker. ``/admin/tool-policy``
already exposes this data, but it is ``admin``-only (INV-5), so an ordinary assistant
owner could not see which tools exist, how risky each one is, or whether their admin
had switched one off. The builder therefore shipped a hardcoded four-entry stub —
which is why ``run_python`` was ungrantable through the UI and the one assistant that
had it was seeded straight into the database.

This endpoint **grants nothing and writes nothing**. It is a read: the registry's
static metadata (``description`` / ``risk_tier`` / ``read_only``) merged with the
tenant's effective ``enabled`` / ``requires_approval`` flags. Every governance
control is untouched — the per-tenant tool policy stays an admin write on
``/admin/tool-policy``, the runner's approval gate still decides whether an allowed
tool may actually run, and the assistant service still validates every allow-list
entry against the registry (an unknown name is a 422). All this does is let the UI
tell the truth about those gates instead of offering a control that silently fails.

Contract-first: the shape matches ``contracts/openapi.yaml`` (``ToolCatalogEntry`` /
``ToolCatalog``). The router validates in → calls **one** service
(:class:`~app.services.tool_policy_service.ToolPolicyService`) → shapes out
(ADR-0004). Bearer-protected via ``current_user`` (a missing/invalid token is a 401,
INV-4) and tenant-scoped via ``current_tenant`` (INV-1) — never request input.
"""

from __future__ import annotations

from fastapi import APIRouter
from pydantic import BaseModel

from app.api.deps import CurrentTenant, CurrentUser, DbSession
from app.services.tool_policy_service import ToolPolicyEntryView, ToolPolicyService

router = APIRouter(tags=["tools"])


# --- Wire models (mirror contracts/openapi.yaml) -----------------------------


class ToolCatalogEntryResponse(BaseModel):
    """``#/components/schemas/ToolCatalogEntry`` — one grantable tool."""

    model_config = {"extra": "forbid"}

    tool_name: str
    description: str
    risk_tier: str
    read_only: bool
    enabled: bool
    requires_approval: bool
    is_default: bool


class ToolCatalogResponse(BaseModel):
    """``#/components/schemas/ToolCatalog`` — every registered tool, in name order."""

    model_config = {"extra": "forbid"}

    items: list[ToolCatalogEntryResponse]


def _to_wire(entries: list[ToolPolicyEntryView]) -> ToolCatalogResponse:
    """Project the service views to the contract wire shape."""
    return ToolCatalogResponse(
        items=[
            ToolCatalogEntryResponse(
                tool_name=entry.tool_name,
                description=entry.description,
                risk_tier=entry.risk_tier,
                read_only=entry.read_only,
                enabled=entry.enabled,
                requires_approval=entry.requires_approval,
                is_default=entry.is_default,
            )
            for entry in entries
        ]
    )


# --- Routes ------------------------------------------------------------------


@router.get("/tools", response_model=ToolCatalogResponse)
async def list_tools(
    _principal: CurrentUser,
    session: DbSession,
    tenant_id: CurrentTenant,
) -> ToolCatalogResponse:
    """Every registered tool with its risk tier and this tenant's effective policy.

    The assistant builder's source for its tool allow-list picker (issue #505): the
    registry is the source of truth for *which* tools exist and their static
    ``risk_tier``/``read_only``/``description``; the tenant's admin overrides (else
    the tool's built-in defaults) supply ``enabled``/``requires_approval`` so the UI
    can show a tool the admin disabled as **unavailable** rather than offering a
    checkbox whose selection would be refused at run time.

    Read-only and grants nothing — the governance write stays admin-only on
    ``PATCH /admin/tool-policy``, and the approval gate is unchanged. Requires a
    valid bearer token (401 otherwise, INV-4); tenant-scoped (INV-1).
    """
    service = ToolPolicyService(session, tenant_id=tenant_id)
    return _to_wire(await service.list_policy())
